lib/display/r_raster.c

```diff
@@ -78,6 +78,9 @@ static void init(void)
   Default display driver is Cairo, if not available PNG is used.
 
   \return 0 on success
+  \return -1 if GRASS_REGION is not defined; use this return value to decide
+  what to do when rendering for the entire computational region can be slow
+  without constraining to the display extent in GRASS_REGION from the monitor
```

Comment:
It's actually both. It can be slow and renders twice. It's rendered the first time in the same file, which will be overwritten the second time. You can see that first render using this patch to the monitor `render.py`:
```patch
--- MONITORS/wx0/render.py.orig 2024-03-14 01:46:02.172981610 -0600
+++ MONITORS/wx0/render.py      2024-03-14 01:45:37.172088784 -0600
@@ -2,6 +2,7 @@
 import os
 import sys
 import tempfile
+import shutil
 
 from grass.script import core as grass
 from grass.script import task as gtask
@@ -163,6 +164,7 @@
     read_stdin(cmd)
 
     render(cmd, mapfile)
+    shutil.copy(mapfile, "/tmp/first_render.ppm");
     update_cmd_file(os.path.join(path, "cmd"), cmd, mapfile)
     if cmd[0] == "d.erase" and os.path.exists(legfile):
         os.remove(legfile)
```
and run `d.rast nidp` for a small region. Check `/tmp/first_render.ppm` to see the entire Texas, not the small region.